piston extruder 5339

BillStruve opened the pull request:
##Why (short version)
Printing by extruding a paste with a stepper motor driven piston does not require
temperature control of the extruder head or the bed.

It does require control of a fan to dry the paste after extrusion.
When not dried, the print will collapse when it is tall.
The manufacturer provides a warm air blower fan for this.
	M35 turns fan on, M36 turns fan off.
	
It also requires a means of adjusting the extrude rate during printing.
The manufacturer has provided a pot with a knob for the user.
	M37 A# - turns on/off the E multiplier Adjustment
	M37 S# - turns on/off Showing of the E multiplier
	M37 R##.# - adjusts the sensitivity of the knob (%)
		equation: y = mx + b  where y = multiplier,
		m = R/100/512,  x = ADC reading, b is calculated
		assuming (max x)/2 = 512 and at the midpoint of the
		adjustment, the multiplier will be 1.000
	
##What (overview)
Configuration.h
	add a section for a piston extruder printer
	add 2 defines to direct compiler
	add 2 default values
Marlin.h
	add 6 extern vars for use in Marlin_main.cpp,
		planner.cpp, and temperature.cpp
Marlin_main.cpp
	set initial values for vars
	add descriptions for M35, M36, and M37
	add code for M35, M36, and M37
	add case's for M35, M36, and M37
enum.h
	add a Prepare/Measure pair for ADC
pins_RAMPS.h
	define a pin for the dryer fan
	define a pin for the ADC input
planner.cpp
	alter extrusion with E multiplier
temperature.cpp
	setup for analog read
	use analog read to calculate E multiplier
	provide user feedback of E multiplier
example_configurations
	add MMM folder
	add a specific Configuration.h to MMM folder

BillStruve committed:
piston extruder 5339
  add a Prepare/Measure pair for ADC

BillStruve committed:
piston extruder 5339
  add MMM folder
  add a specific Configuration.h to MMM

BillStruve committed:
piston extruder 5339
  add a section for a piston extruder printer
  add 2 defines to direct compiler
  add 2 default values

BillStruve committed:
piston extruder 5339
  add 6 extern vars for use in Marlin_main.cpp, planner.cpp, and temperature.cpp

BillStruve committed:
piston extruder 5339
  set initial values for vars
  add descriptions for M35, M36, and M37
  add code for M35, M36, and M37
  add case's for M35, M36, and M37

BillStruve committed:
piston extruder 5339
  define a pin for the dryer fan
  define a pin for the ADC input

BillStruve committed:
piston extruder 5339
  alter extrusion with E multiplier

BillStruve committed:
piston extruder 5339
  setup for analog read
  use analog read to calculate E multiplier
  provide user feedback of E multiplier

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -81,6 +81,22 @@
+//
+
+//#define PISTON_EXTRUDER //used to set parameters for the Piston Extruder
+#ifdef PISTON_EXTRUDER
```
Please use `#if ENABLED(...)` for options like these.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4810,6 +4825,65 @@ inline void gcode_M31() {
 
 #endif // SDSUPPORT
 
+#ifdef PISTON_EXTRUDER
```
`#if ENABLED(PISTON_EXTRUDER)`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4810,6 +4825,65 @@ inline void gcode_M31() {
 
+#ifdef PISTON_EXTRUDER
+  inline void gcode_M35() { //turn on dryer fan
+    WRITE(DRYER_FAN_PIN, 1);
```
Please use `HIGH` instead of `1`.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4810,6 +4825,65 @@ inline void gcode_M31() {
+    SERIAL_ECHOLN("Dryer fan ON");
+  }  
+  inline void gcode_M36() { //turn off dryer fan
+    WRITE(DRYER_FAN_PIN, 0);
```
Please use `LOW` rather than `0`.

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -162,7 +162,17 @@
 #endif
 
-#define HEATER_0_PIN     RAMPS_D10_PIN
+#ifdef PISTON_EXTRUDER
```
`#if ENABLED(PISTON_EXTRUDER)`

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -162,7 +162,17 @@
+#ifdef PISTON_EXTRUDER
+  #define DRYER_FAN_PIN  RAMPS_D10_PIN //to control the dryer fan
+  #define HEATER_0_PIN   RAMPS_D9_PIN  //needed to pass sanity check
+  #ifdef E_ADJUST_MANUALLY
```
`#if ENABLED(E_ADJUST_MANUALLY)`

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -162,7 +162,17 @@
+  #endif
+#else
+  #define HEATER_0_PIN   RAMPS_D10_PIN
+#endif  
```
And... Actually, when defining pins for a specific function in the pins files, don't set conditions on them. We're defining the default pins that _will be used_ when the feature that uses them is enabled. When the feature is disabled, we just don't touch the pins.

thinkyhead commented on Marlin/planner.cpp:
```diff
@@ -755,7 +755,12 @@ void Planner::_buffer_line(const float &a, const float &b, const float &c, const
+    float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * e_adjust_multiplier * 0.01;
+  #else
+    float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * 0.01;
+  #endif
```
Marlin code is leaning towards this style, to reduce repetition:
```cpp
float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * 0.01
  #if ENABLED(E_ADJUST_MANUALLY)
    * e_adjust_multiplier
  #endif
;
```

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1809,13 +1812,41 @@ void Temperature::isr() {
-        raw_temp_value[3] += ADC;
-      #endif
-      temp_state = Prepare_FILWIDTH;
-      break;
```
Here I suggest keeping `MeasureTemp3` mostly as it was, but just override the new state it sets:
```cpp
temp_state =
  #if ENABLED(E_ADJUST_MANUALLY)
    Prepare_e_adjust_reading
  #else
    Prepare_FILWIDTH
  #endif
;
```

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1809,13 +1812,41 @@ void Temperature::isr() {
+        temp_state = Prepare_e_adjust_reading;
+        break;
 
+      case Prepare_e_adjust_reading:
```
…and then you can have `#if ENABLED(E_ADJUST_MANUALLY)` just above this line ^

thinkyhead reviewed (changes requested):
Looking pretty good. Patch up with the requested changes, rebase, squash, and `git push -f` the updated branch.

BillStruve commented:
Made all changes. 
Did a git rebase -i HEAD~1 since all changes were made in one add/commit, did not need to squash
Did a git push -f https://github.com/BillStruve/PistonExtruder.git
Hope that works.

BillStruve committed:
Requested changes

BillStruve committed:
fixed typos

BillStruve committed:
commented out #define Piston in Configuration.h

BillStruve committed:
fixed Configuration.h

BillStruve commented:
Does anything else need to be done?

BillStruve committed:
Merge remote-tracking branch 'refs/remotes/RCBugFix/RCBugFix' into RCBugFix

BillStruve committed:
piston extruder 5339
  add a section for a piston extruder printer
  add 2 defines to direct compiler
  add 2 default values

BillStruve committed:
Added _MAX_ to ABL_GRID_MAX_POINTS

BillStruve commented:
Updated to make E adjust independent of dryer fan.
